File: Kotlin/Maths/ArmstrongRecursive.kt
/*Kotlin Program to check Given number is an Armstrong Number or not*/
import java.util.*

// Function for check Armstrong Number
fun isArmstrongNumber(num: Int, size: Int): Int {
    if (num<10) {
        return Math.pow(num.toDouble(), size.toDouble()).toInt()
    } else {
        return(Math.pow((num % 10).toDouble(), size.toDouble()).toInt() + isArmstrongNumber(num / 10, size))
    }
}

// Main Function, Entry Point of Program
fun main(arg: Array<String>) {
    val sc = Scanner(System.`in`)

    println("Enter Number: ")
    val num: Int = sc.nextInt()
    val size: Int = Integer.toString(num).length

    // Call function to check number is Armstrong
    if (num == isArmstrongNumber(num, size)) {
        println("$num is an Armstrong Number")
    } else {
        println("$num is not an Armstrong Number")
    }
}

/**
 * Sample input/output:
 * 
 * Enter Number: 
 * 153
 * 153 is an Armstrong Number
 * 
 * Enter Number: 
 * 224
 * 224 is not an Armstrong Number
 */

File: Dart/maths/ArmstrongRecursive.dart
/**
This function is a recursive function, which calls itself over and over
according to the number of digits in the number. Using that and adding each
digit raised to the power of the number of digits, it keeps adding them
and we get the sum we need to compare with the number input.
 */
import 'dart:io';
import 'dart:math';

// recursive function which add digits raised to the power of number of digits
int armstrong(int num, int len) {
  if(num < 10) {
    return pow(num, len).toInt();
  } else {
    return (pow(num % 10, len).toInt() + armstrong(num ~/ 10, len));
  }
}

// Main Function, Entry Point of Program
void main() {
  print("Enter a number:");
  int num = int.parse(stdin.readLineSync()!);
  int len = num.toString().length;

  // Call function to check number is Armstrong
  if (num == armstrong(num, len)) {
    print("$num is an Armstrong Number");
  } else {
    print("$num is not an Armstrong Number");
  }
}

/**
Sample Input and Output :
Enter a number:
153
153 is an Armstrong Number

Enter a number:
213
213 is not an Armstrong Number

Time complexity = O(n)
Space complexity = O(1)
 */
